File: apps/web/src/app/schemas/tagFrame.ts
/** Mirror of packages/contracts/tag_frame.schema.json */

import type { TagQuality, TagSource, TagValue } from "./common";

export interface TagFrame {
  tag_id: string;
  asset_id: string;
  value: TagValue;
  unit: string;
  quality: TagQuality;
  /** RFC3339 UTC */
  timestamp: string;
  source: TagSource;
  seq?: number;
  /** RFC3339 UTC */
  ingest_ts?: string;
  gateway_id?: string;
  scenario_id?: string;
}

/** Dedupe key per contract: (source, tag_id, seq, timestamp). */
export function tagFrameIdentityKey(frame: TagFrame): string {
  return `${frame.source}:${frame.tag_id}:${frame.seq ?? ""}:${frame.timestamp}`;
}

File: apps/api/app/schemas/tag_frame.py
"""Mirror of packages/contracts/tag_frame.schema.json."""

from datetime import datetime

from pydantic import AwareDatetime, BaseModel, ConfigDict, Field

from app.schemas.common import TagQuality, TagSource, TagValue


class TagFrame(BaseModel):
    """Universal live telemetry envelope (simulator and gateway emit identical shape)."""

    model_config = ConfigDict(extra="forbid")

    tag_id: str = Field(pattern=r"^[A-Z0-9_]+$")
    asset_id: str = Field(pattern=r"^[A-Z0-9-]+$")
    value: TagValue
    unit: str
    quality: TagQuality
    timestamp: AwareDatetime
    source: TagSource
    seq: int | None = Field(default=None, ge=0)
    ingest_ts: AwareDatetime | None = None
    gateway_id: str | None = None
    scenario_id: str | None = None

    def identity_key(self) -> tuple[str, str, int | None, datetime]:
        """Dedupe key per contract: (source, tag_id, seq, timestamp)."""
        return (self.source, self.tag_id, self.seq, self.timestamp)